### PR TITLE
fix golem marker loading failure

### PR DESCRIPTION
**Pull request by albus12138**
Tested in 2.3.0
Fixes https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/11829
Fixes https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/8942

### CODE REVIEW

**Comment by TimeConqueror** [comment]
Why do you use `@Overwrite` instead of more readable and handleable `@Inject`-based modification?

**Comment by mitchej123** [comment]
> Why do you use `@Overwrite` instead of more readable and handleable `@Inject`-based modification?

+1. Would prefer a non overwrite solution; it's a last resort generally and isn't generally compatible with other mixins/asm.

**Comment by albus12138** [comment]
I have replace @Overwrite with @Inject in EntityGolemBase and it works. But when I tried to inject into ItemGolemBell, I got this error:

```
cpw.mods.fml.common.LoaderException: java.lang.NoClassDefFoundError: com/kentington/thaumichorizons/common/items/ItemGolemBellTH
    at cpw.mods.fml.common.LoadController.transition(LoadController.java:163)
    at cpw.mods.fml.common.Loader.loadMods(Loader.java:544)
    at cpw.mods.fml.client.FMLClientHandler.beginMinecraftLoading(FMLClientHandler.java:208)
    at net.minecraft.client.Minecraft.func_71384_a(Minecraft.java:480)
    at net.minecraft.client.Minecraft.func_99999_d(Minecraft.java:5093)
    at net.minecraft.client.main.Main.main(SourceFile:148)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:568)
    at net.minecraft.launchwrapper.Launch.launch(Launch.java:165)
    at net.minecraft.launchwrapper.Launch.main(Launch.java:29)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:568)
    at org.multimc.onesix.OneSixLauncher.launchWithMainClass(OneSixLauncher.java:234)
    at org.multimc.onesix.OneSixLauncher.launch(OneSixLauncher.java:269)
    at org.multimc.EntryPoint.listen(EntryPoint.java:143)
    at org.multimc.EntryPoint.main(EntryPoint.java:34)
Caused by: java.lang.NoClassDefFoundError: com/kentington/thaumichorizons/common/items/ItemGolemBellTH
    at java.lang.Class.forName0(Native Method)
    at java.lang.Class.forName(Class.java:467)
    at cpw.mods.fml.common.FMLModContainer.constructMod(FMLModContainer.java:440)
    at jdk.internal.reflect.GeneratedMethodAccessor9.invoke(Unknown Source)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:568)
    at com.google.common.eventbus.EventSubscriber.handleEvent(EventSubscriber.java:74)
    at com.google.common.eventbus.SynchronizedEventSubscriber.handleEvent(SynchronizedEventSubscriber.java:47)
    at com.google.common.eventbus.EventBus.dispatch(EventBus.java:322)
    at com.google.common.eventbus.EventBus.dispatchQueuedEvents(EventBus.java:304)
    at com.google.common.eventbus.EventBus.post(EventBus.java:275)
    at cpw.mods.fml.common.LoadController.sendEventToModContainer(LoadController.java:212)
    at cpw.mods.fml.common.LoadController.propogateStateMessage(LoadController.java:190)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:568)
    at com.google.common.eventbus.EventSubscriber.handleEvent(EventSubscriber.java:74)
    at com.google.common.eventbus.SynchronizedEventSubscriber.handleEvent(SynchronizedEventSubscriber.java:47)
    at com.google.common.eventbus.EventBus.dispatch(EventBus.java:322)
    at com.google.common.eventbus.EventBus.dispatchQueuedEvents(EventBus.java:304)
    at com.google.common.eventbus.EventBus.post(EventBus.java:275)
    at cpw.mods.fml.common.LoadController.distributeStateMessage(LoadController.java:119)
    at cpw.mods.fml.common.Loader.loadMods(Loader.java:513)
    ... 18 more
Caused by: java.lang.ClassNotFoundException: com.kentington.thaumichorizons.common.items.ItemGolemBellTH
    at net.minecraft.launchwrapper.LaunchClassLoader.findClass(LaunchClassLoader.java:212)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:587)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:520)
    ... 42 more
Caused by: java.lang.NoClassDefFoundError: thaumcraft/common/entities/golems/ItemGolemBell
    at java.lang.ClassLoader.defineClass1(Native Method)
    at java.lang.ClassLoader.defineClass(ClassLoader.java:1012)
    at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:150)
    at net.minecraft.launchwrapper.LaunchClassLoader.findClass(LaunchClassLoader.java:202)
    ... 44 more
Caused by: java.lang.ClassNotFoundException: thaumcraft.common.entities.golems.ItemGolemBell
    at net.minecraft.launchwrapper.LaunchClassLoader.findClass(LaunchClassLoader.java:114)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:587)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:520)
    ... 48 more
```

Here is the code:
```
@Mixin(ItemGolemBell.class)
public class MixinItemGolemBell {

    @Inject(method = "getMarkers", at = @At(value = "TAIL"), remap = false)
    public void hodgepodge$getMarkers(ItemStack stack, CallbackInfoReturnable<ArrayList<Marker>> cir) {
        ArrayList<Marker> markers = cir.getReturnValue();
        NBTTagList nbttaglist = stack.stackTagCompound.getTagList("markers", 10);
        for (int i = 0; i < nbttaglist.tagCount(); i++) {
            NBTTagCompound nbttagcompound1 = nbttaglist.getCompoundTagAt(i);
            Marker marker = markers.get(i);
            int dim = nbttagcompound1.getInteger("dim");
            if (marker.dim != dim) {
                marker.dim = dim;
            }
        }
        cir.setReturnValue(markers);
    }

}
```

How can I fix it?

File: src/main/java/com/mitchej123/hodgepodge/mixins/late/thaumcraft/MixinEntityGolemBase.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    @Inject(method = "readEntityFromNBT", at = @At(value = "TAIL"))
+    public void hodgepodge$readEntityFromNBT(NBTTagCompound nbt, CallbackInfo ci) {
+        NBTTagList nbttaglist = nbt.getTagList("Markers", 10);
+        for (int i = 0; i < nbttaglist.tagCount(); i++) {
+            NBTTagCompound nbttagcompound1 = nbttaglist.getCompoundTagAt(i);
+            Marker marker = this.markers.get(i);
+            int dim = nbttagcompound1.getInteger("dim");
+            if (marker.dim != dim) {
```
Comment (TimeConqueror) [inline src/main/java/com/mitchej123/hodgepodge/mixins/late/thaumcraft/MixinEntityGolemBase.java:36]: if-statement is not needed here

File: src/main/java/com/mitchej123/hodgepodge/mixins/late/thaumcraft/MixinItemGolemBell.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+    @Inject(method = "getMarkers", at = @At(value = "TAIL"), remap = false, cancellable = true)
+    private static void hodgepodge$getMarkers(ItemStack stack, CallbackInfoReturnable<ArrayList<Marker>> cir) {
+        ArrayList<Marker> markers = cir.getReturnValue();
+        NBTTagList nbttaglist = stack.stackTagCompound.getTagList("markers", 10);
+        for (int i = 0; i < nbttaglist.tagCount(); i++) {
+            NBTTagCompound nbttagcompound1 = nbttaglist.getCompoundTagAt(i);
+            Marker marker = markers.get(i);
+            int dim = nbttagcompound1.getInteger("dim");
+            if (marker.dim != dim) {
```
Comment (TimeConqueror) [inline src/main/java/com/mitchej123/hodgepodge/mixins/late/thaumcraft/MixinItemGolemBell.java:28]: same

**Comment by TimeConqueror** [comment]
One more thing: I'd move the common part into a separate method, if it's possible

**Review by TimeConqueror — APPROVED** [review]

**Comment by Alastors** [comment]
Shouldn't this go in TC4Tweaks like the rest of thaum 4 mixin fixes?

**Review by Alastors — CHANGES_REQUESTED** [review]
Move to TC4Tweaks to maintain consistency

**Comment by albus12138** [comment]
@Alastors There is no mixin in TC4Tweaks (all of them are asm fixes), but I know nothing about ASM.

Hodgepodge has both asm and mixin and its goal is to fix other mods in GTNH. TC4Tweaks only works for Thaumcraft 4. For maintain consistency, I think move tc4tweaks into hodgepodge is more reasonable.

**Comment by miozune** [comment]
The reason TC4Tweaks not using Mixin is it doesn't require external library mod, which is good for compat with outside of GTNH. However, since ASM is harder to write than Mixin, keeping this fix in Hodgepodge would be acceptable, than not having the fix at all due to no one working on porting to ASM.

**Review by mitchej123 — APPROVED** [review]

**Comment by Alastors** [comment]
It's absolutely not reasonable to move TC4Tweaks into Hodgepodge when the former is meant to work outside the pack

**Comment by mitchej123** [comment]
I mean, hodgepodge works "outside the pack" reasonably well.  Tc4tweaks currently has no mixins, and rewriting this as ASM is non trivial for most people.  I think it's fine to ship with hodgepodge for now, and if glee8e would like to move it to tc4tweaks that's great.

**Pull request merged**